### PR TITLE
[Feature][LoRA] 1.Support W8A8 dynamic multi-LoRA inference for MoE models; 2. support shared expert + lora;3. Gereneral quant MOE FrameWork for Lora inference 

### PR DESCRIPTION
**Pull request by SkychenLee**
## Motivation

This PR adds Ascend support for multi-LoRA inference on MoE models with a
W8A8_DYNAMIC quantized base model.

The main target is DeepSeek V4 Flash-style models containing routed experts and
shared experts. LoRA computation must preserve floating-point activation
boundaries while the base expert GMMs remain quantized.

## Verification
Here I verified this pr from deepseekv4 flash w8a8-int8, and the lora is trainied by mindspeed.
Since mindspeed does not support transfering dsv4flash lora checkpoint  from megatron to peft, so I solve this by my self https://github.com/SkychenLee/LoraAdapterTransformer/blob/main/README.md , you  can  directly use the scritpt to transfer the lora adapter for dsv4 flash or you can use the skill I provided to transfer any model's lora adapters from any code agent.
Here are the command for **start up** of the model
```
#!/usr/bin/bash
export OMP_PROC_BIND=false
export OMP_NUM_THREADS=10
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True

export TASK_QUEUE_ENABLE=1
export HCCL_OP_EXPANSION_MODE="AIV"
export HCCL_BUFFSIZE=512

sysctl -w vm.swappiness=0
sysctl -w kernel.numa_balancing=0
sysctl kernel.sched_migration_cost_ns=50000
 
vllm serve "your deep seek v4 flash moel path" \
  --safetensors-load-strategy 'prefetch' \
  --max-model-len 10240 \
  --max-num-batched-tokens 10240 \
  --served-model-name ds \
  --gpu-memory-utilization 0.9 \
  --max-num-seqs 32 \
  --data-parallel-size 1 \
  --tensor-parallel-size 8 \
  --quantization ascend \
  --port 7000 \
  --block-size 128 \
  --enable-chunked-prefill \
  --no-enable-prefix-caching \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --reasoning-parser deepseek_v4 \
  --async-scheduling \
  --enable-lora \
  --lora-modules lora1="your lora adapter path"\
  --max-lora-rank 16 
```
### Routed expert LoRA

- Add Ascend-native routed expert LoRA execution for vLLM v0.23.0.
- Propagate the per-layer LoRA context through MoE dispatch and MLP execution.
- Recover the `(expert_id, LoRA slot)` mapping after expert routing to support:
  - multiple adapters in one batch;
  - base and LoRA requests mixed in one batch;
  - different adapters selected by different tokens.
- Apply:
  - `w13` LoRA after GMM1 and before activation;
  - `w2` LoRA after GMM2;
  - base tokens with zero LoRA delta.
- Add separate `w13` and `w2` adapter enable masks so adapters containing only
  part of the MoE projections do not execute unintended deltas.

### Quantized MoE LoRA

- Add an extensible quantized MoE LoRA implementation registry keyed by
  `QuantType`.
- Keep the public execution and dispatch interfaces independent of W8A8.
- Let each quantization implementation define:
  - whether dispatch-side activation quantization must be disabled;
  - how dispatch inputs should be validated;
  - how the quantized expert MLP should be executed.
- Register the first implementation for W8A8_DYNAMIC:
  1. preserve the routed BF16/FP16 input;
  2. dynamically quantize input for INT8 GMM1;
  3. inject BF16 `w13` LoRA;
  4. run SwiGLU/activation;
  5. dynamically quantize activation for INT8 GMM2;
  6. inject BF16 `w2` LoRA.
- Keep the existing fused W8A8 path unchanged for base-only batches.
- Refresh `no_lora` metadata during decode so base-only decode batches do not
  unnecessarily enter the LoRA path.
- Fail explicitly for quantization schemes without a registered MoE LoRA
  implementation.

### Shared expert LoRA

- Preserve the `_shared_experts` module hierarchy so shared-expert adapter
  weights can be discovered and loaded by the vLLM LoRA model manager.
- Execute shared experts through the existing dense LoRA wrappers when LoRA is
  active.
- Unwrap dense LoRA wrappers only for the optimized base-only quantized shared
  expert path.
- Add a gather+BMM expand-slice fallback for shared-expert dense LoRA B
  projections that are incompatible with the existing NPU expand-slice shape
  restrictions.
- Automatically enable the compatible expand-slice path when shared experts
  are present; no new user-facing option is introduced.

## Supported scope

The initial quantized implementation supports:

- W8A8_DYNAMIC base experts;
- AllGather communication;
- tensor parallelism without expert parallelism;
- statically loaded multiple adapters;
- mixed base and multi-adapter requests in the same batch;
- routed experts, shared experts, and existing attention LoRA paths.

The following combinations are not supported by this initial implementation
and fail explicitly:

- expert parallelism and All-to-All;
- MC2/FusedMC2;
- dynamic EPLB;
- derived W8A8FP8 and PDMix MoE schemes;
- multistream-overlap gate execution.

Other quantization formats can be added through the new quantized MoE LoRA
registry without changing the generic token dispatcher or MoE MLP entry.

## Tests

Unit tests were added for:

- LoRA-active and base-only dispatch quantization behavior;
- single-adapter, multi-adapter, and base/adapter token routing;
- routed expert mapping after token sorting;
- BF16 LoRA injection around the two quantized GMM operations;
- two dynamic activation quantization calls;
- independent `w13` and `w2` adapter enable masks;
- shared-expert discovery and dense LoRA execution;
- gather+BMM expand-slice fallback;
- TP LoRA weight slicing;
- unsupported execution modes and unregistered quantization formats.

Static validation completed:

- Python compilation checks;
- Ruff lint and formatting checks;
- Git whitespace checks.

## Compatibility

- No new CLI arguments or request fields are introduced.
- The existing PEFT adapter format is unchanged.
- BF16 MoE LoRA continues to use the existing unquantized execution path.
- Base-only W8A8 batches retain the original fused quantized path.

- vLLM version: v0.23.0
- vLLM main: https://github.com/vllm-project/vllm/commit/ee0da84ab9e04ac7610e28580af62c365e898389

### CODE REVIEW

File: vllm_ascend/lora/punica_npu.py [hunk 1]
```diff
@@ -9,6 +9,57 @@
...
+
+    if x_active.shape[0] == 0 or gathered.shape[1] == 0:
+        return
+
+    delta = torch.bmm(gathered, x_active.unsqueeze(-1)).squeeze(-1)
+    delta = torch.where(
+        (indices >= 0).unsqueeze(-1),
+        delta,
+        torch.zeros_like(delta),
+    )
```
[the remark this answers is not in the document]
Comment (SkychenLee) [reply vllm_ascend/lora/punica_npu.py]: I have replace the code with masked_fill_
